parsers/CA_NS.py

```diff
@@ -1,212 +1,175 @@
     logger: Logger = getLogger(__name__),
-) -> list[dict]:
+) -> list[dict[str, Any]]:
     """Requests the last known production mix (in MW) of a given country."""
     if target_datetime:
-        raise NotImplementedError(
-            "This parser is unable to give information more than 24 hours in the past"
-        )
+        raise ParserException(PARSER, "Unable to fetch historical data", zone_key)
+
+    if zone_key != ZONE_KEY:
+        raise ParserException(PARSER, f"Cannot parse zone '{zone_key}'", zone_key)
 
-    r = session or Session()
+    # Request data from the source. Skip the first element of each JSON array
+    # because the reported base load is always 0 MW.
+    session = session or Session()
+    loads = {  # A lookup table mapping timestamps to base loads (in MW)
+        _parse_timestamp(load["datetime"]): load["Base Load"]
+        for load in session.get(LOAD_URL).json()[1:]
+    }
+    mixes = session.get(MIX_URL).json()[1:]  # Electricity mix breakdowns in %
+
+    production_breakdowns = ProductionBreakdownList(logger)
+    for mix in mixes:
+        timestamp = _parse_timestamp(mix["datetime"])
+        if timestamp not in loads:
+            logger.warning(
+                "unable to find base load for %s", timestamp, extra={"key": ZONE_KEY}
+            )
+            continue
+        load = loads[timestamp]
+        if load <= 0:
+            logger.warning("invalid base load of %s MW", load, extra={"key": ZONE_KEY})
+            continue
 
-    production, imports = _get_ns_info(r, logger)
+        production_mix = ProductionMix()
+        production_mix.add_value("biomass", load * mix["Biomass"] / 100)
+        production_mix.add_value("coal", load * mix["Solid Fuel"] / 100)
+        production_mix.add_value("gas", load * mix["HFO/Natural Gas"] / 100)
+        production_mix.add_value("gas", load * mix["LM 6000's"] / 100)
+        production_mix.add_value("hydro", load * mix["Hydro"] / 100)
+        production_mix.add_value("oil", load * mix["CT's"] / 100)
+        production_mix.add_value("wind", load * mix["Wind"] / 100)
+        # Sanity checks: verify that reported production doesn't exceed listed
+        # capacity by a lot. In particular, we've seen error cases where hydro
+        # production ends up calculated as 900 MW which greatly exceeds known
+        # capacity of around 520 MW.
+        if (
+            100 < (production_mix.biomass or 0)
+            or 1300 < (production_mix.coal or 0)
+            or 700 < (production_mix.gas or 0)
+            or 600 < (production_mix.hydro or 0)
+            or 300 < (production_mix.oil or 0)
+            or 700 < (production_mix.wind or 0)
+        ):
+            logger.warning(
+                "discarding datapoint at %s because some mode's production is "
+                "infeasible: %s",
+                timestamp,
+                production_mix,
+                extra={"key": ZONE_KEY},
+            )
+            continue
```

Comment:
@mathilde-daugy do we still want to do this at a parser level or should we do it with the new validation layer in the backend?

Reply:
the validation layer is not yet ready so we can keep that dor now and update when relevant